Web Inspector: Update esprima-next

rcaliman-apple opened the pull request:
#### 2c033825b783f71f483e47c70f4d1398b40feaa7
<pre>
Web Inspector: Update esprima-next
<a href="https://bugs.webkit.org/show_bug.cgi?id=300387">https://bugs.webkit.org/show_bug.cgi?id=300387</a>
<a href="https://rdar.apple.com/162193882">rdar://162193882</a>

Reviewed by Devin Rousso.

Update esprima-next to the current latest version
(<a href="https://github.com/node-projects/esprima-next/commit/242fb79474f623c0efba35a25fa96ea48e0db35d)">https://github.com/node-projects/esprima-next/commit/242fb79474f623c0efba35a25fa96ea48e0db35d)</a>
which correctly identifes `for` statements that contain optional chaining operators.
This enables JSFormatter.js to add whitespace in the right places for these cases when pretty printing.

Add a new script that can update WebKit&apos;s exprima-next checkout
to the latest version with a configuration compatible with Web Inspector.

The default bundler configuration in the esprima-next source code outputs an ES module.
Web Inspector cannot currently use this. The script contains a custom bundler configuration
which produces an output with `esprima` exposed to the global scope, as Web Inspector expects it.

* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements-expected.js:
(c.d.1):

Rebaseline tests for optional chaining operators in `for` statements added in <a href="https://commits.webkit.org/301197@main">https://commits.webkit.org/301197@main</a>

* Source/WebInspectorUI/Scripts/update-esprima-next: Added.
* Source/WebInspectorUI/UserInterface/External/Esprima/esprima.js:
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: <a href="https://commits.webkit.org/301430@main">https://commits.webkit.org/301430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/552d1f3684f76c2cfd86cdf35bf39fca8e10d355

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45601 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36369 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132809 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77798 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aafa7ba7-0b75-4c4c-a358-3d5ae62cdb81) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54159 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95951 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64044 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/407ec145-be79-44d0-b558-370f1a2a46d6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128888 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112624 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76440 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f29fa904-881c-4daf-b891-735955863192) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/125292 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35925 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30808 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76279 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106799 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31020 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135494 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52724 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40455 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104423 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53173 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108833 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104143 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26524 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49519 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27840 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50102 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52618 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58431 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51964 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55311 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53660 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/Scripts/update-esprima-next:
```diff
@@ -0,0 +1,120 @@
+echo "✓ Changed directory into: $(pwd)"
+echo "Cloning repository: $REPO_URL"
+
+if git clone "$REPO_URL"; then
```
should we add this to the `.gitignore` so that it's not accidentally committed if the developer doesn't remove the repo

rcaliman-apple replied:
I was hesitant to add that, since then it would be easy to forget an old copy of the esprima-next repo in the local source (because it doesn't show up in `git status`. But combined with the `git pull` suggestion from the other comment, these two work well together to mitigate old checkouts. 👍 Added.

dcrousso commented on Source/WebInspectorUI/Scripts/update-esprima-next:
```diff
@@ -0,0 +1,120 @@
+    echo "✓ Repository cloned to: $(pwd)/$REPO_NAME"
+else
+    if [ -d "$REPO_NAME" ] && [ -d "$REPO_NAME/.git" ]; then
+        echo "Warning: A git repository already exists in '$REPO_NAME'"
```
NIT: if the repo does exist should we `git pull`?

dcrousso commented on Source/WebInspectorUI/Scripts/update-esprima-next:
```diff
@@ -0,0 +1,120 @@
+
+NEW_ESPRIMA="$(pwd)/dist/esprima.js"
+
+npx tsup --config $TSUP_CONFIG_FILENAME;
```
might be worth mentioning/checking earlier that this script requires `node` and/or `npm` to be installed beforehand

rcaliman-apple replied:
Good idea! Added checks.

dcrousso reviewed (approved):
nice!  love all of those `for`matting improvements 😜

rcaliman-apple committed:
Web Inspector: Update esprima-next
https://bugs.webkit.org/show_bug.cgi?id=300387
rdar://162193882

Reviewed by Devin Rousso.

Update esprima-next to the current latest version
(https://github.com/node-projects/esprima-next/commit/242fb79474f623c0efba35a25fa96ea48e0db35d)
which correctly identifes `for` statements that contain optional chaining operators.
This enables JSFormatter.js to add whitespace in the right places for these cases when pretty printing.

Add a new script that can update WebKit's exprima-next checkout
to the latest version with a configuration compatible with Web Inspector.

The default bundler configuration in the esprima-next source code outputs an ES module.
Web Inspector cannot currently use this. The script contains a custom bundler configuration
which produces an output with `esprima` exposed to the global scope, as Web Inspector expects it.

* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements-expected.js:
(c.d.1):

Rebaseline tests for optional chaining operators in `for` statements added in https://commits.webkit.org/301197@main

* Source/WebInspectorUI/Scripts/update-esprima-next: Added.
* Source/WebInspectorUI/UserInterface/External/Esprima/esprima.js:
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: https://commits.webkit.org/301430@main
